service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+	}); err != nil {
+		return ackLevel, nil
+	}
+
+	return newAckLevelTaskID, nil
+}
+
+func (c *crossClusterQueueProcessorBase) readTasks(
+	readLevel task.Key,
+	maxReadLevel task.Key,
+) ([]*persistence.CrossClusterTaskInfo, bool, error) {
+
+	var response *persistence.GetCrossClusterTasksResponse
+	op := func() error {
+		var err error
+		response, err = c.shard.GetExecutionManager().GetCrossClusterTasks(
+			context.Background(),
+			&persistence.GetCrossClusterTasksRequest{
+				TargetCluster: c.targetCluster,
+				ReadLevel:     readLevel.(transferTaskKey).taskID,
+				MaxReadLevel:  maxReadLevel.(transferTaskKey).taskID,
+				BatchSize:     c.options.BatchSize(),
+			})
+		return err
+	}
+
+	err := backoff.Retry(op, persistenceOperationRetryPolicy, persistence.IsBackgroundTransientError)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return response.Tasks, len(response.NextPageToken) != 0, nil
+}
+
+func (c *crossClusterQueueProcessorBase) pollTasks() []task.Task {
+
+	var result []task.Task
+	for _, queueTask := range c.getTasks() {
+		crossClusterTask, ok := queueTask.(task.CrossClusterTask)
+		if !ok {
+			panic("received non cross cluster task")
+		}
+		if crossClusterTask.IsReadyForPoll() {
+			result = append(result, crossClusterTask)
+		}
+	}
+	return result
+}
+
+func (c *crossClusterQueueProcessorBase) getTask(key task.Key) (task.Task, error) {
+	for _, collection := range c.processingQueueCollections {
+		if task, err := collection.GetTask(key); err == nil {
+			return task, nil
+		}
+	}
+	return nil, errTaskNotFound
+}
+
+func (c *crossClusterQueueProcessorBase) getTasks() []task.Task {
+	var outstandingTask []task.Task
+	for _, collection := range c.processingQueueCollections {
+		outstandingTask = append(outstandingTask, collection.GetTasks()...)
+	}
+	return outstandingTask
+}
+
+func (c *crossClusterQueueProcessorBase) updateTask(
+	taskKey task.Key,
+	remoteResponse interface{},
+) error {
+
+	queueTask, err := c.getTask(taskKey)
+	if err != nil {
+		return err
+	}
+	crossClusterTask, ok := queueTask.(task.CrossClusterTask)
+	if ok {
+		if err := crossClusterTask.Update(remoteResponse); err != nil {
+			c.logger.Error("failed to update cross cluster task",
+				tag.TaskID(crossClusterTask.GetTaskID()),
+				tag.WorkflowDomainID(crossClusterTask.GetDomainID()),
+				tag.WorkflowID(crossClusterTask.GetWorkflowID()),
+				tag.WorkflowRunID(crossClusterTask.GetRunID()),
+				tag.Error(err))
+			return err
+		}
+	} else {
+		panic("Received non cross cluster task.")
+	}
+
+	// if the task fails to submit to main queue, it will use redispatch queue
+	_, err = c.submitTask(crossClusterTask)
+	return err
+}
+
+func (c *crossClusterQueueProcessorBase) addOutstandingTaskCountLocked(count int) {
+	c.outstandingTaskCount += count
+}
+
+// TODO: this will be called in cross cluster task Acked function.
+func (c *crossClusterQueueProcessorBase) removeOutstandingTaskCountLocked(count int) {
+	c.outstandingTaskCount -= count
+}
```

Comment:
`Ack()` method will be called concurrently on many tasks, so do we need a lock on the `outstandingTaskCount` field?

Reply:
I can move the locking logic into this method